File: 11. Linked List/codes/Rewrite(recommended)/singlyll.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

//LinkedList Class
type LinkedList struct {
	headNode *Node
}

func (L *LinkedList) AddToHead(val int) {
	var node = Node{}
	node.data = val
	if L.headNode != nil {
		node.next = L.headNode
	}
	L.headNode = &node
}

func (L *LinkedList) AddToEnd(val int) {
	var node = Node{}
	var temp = L.headNode

	node.data = val
	if temp == nil {
		L.headNode = &node
		return
	}

	for {
		if temp.next == nil {
			temp.next = &node
			break
		}
		temp = temp.next
	}
}

func (L *LinkedList) Display() {
	var temp *Node //OR, &Node{}
	temp = L.headNode
	for temp != nil {
		fmt.Print(temp.data, " ")
		temp = temp.next
	}
}

func main() {
	var ll1 = LinkedList{}
	ll1.AddToHead(3)
	ll1.AddToHead(5)
	ll1.AddToHead(9)
	ll1.AddToEnd(1)
	ll1.AddToEnd(54)
	ll1.AddToHead(0)
	ll1.Display()
}

/* without the LinkedList class
package main

import "fmt"

type Node struct {
	data rune
	next *Node
}

func Display(head *Node) {
	var temp *Node = head
	for temp != nil {
		fmt.Printf("%c", temp.data)
		temp = temp.next
	}
}

func main() {
	var head *Node
	head = &Node{'a', nil}

	var currNode *Node = head

	var i rune
	for i = 'b'; i <= 'z'; i++ {
		currNode.next = &Node{i, nil}
		currNode = (*currNode).next
	}

	Display(head)
}
*/

File: 11. Linked List/codes/Rewrite(recommended)/singlyll.rs
use std::cell::RefCell;
use std::rc::Rc;

type SingleLink = Option<Rc<RefCell<Node>>>;

struct Node {
    value: i32,
    next: SingleLink,
}

struct LinkedList {
    head: SingleLink,
    tail: SingleLink,
    pub length: u64,
}

impl Node {
    fn new(value: i32) -> Rc<RefCell<Node>> {
        Rc::new(RefCell::new(Node {
            value: value,
            next: None,
        }))
    }
}

impl LinkedList {
    fn new_empty() -> LinkedList {
        LinkedList {
            head: None,
            tail: None,
            length: 0,
        }
    }

    pub fn append(&mut self, value: i32) {
        let new_node = Node::new(value);

        match &self.tail {
            Some(v) => {
                v.borrow_mut().next = Some(new_node.clone());
            }
            None => self.head = Some(new_node.clone()),
        };
        self.length += 1;
        self.tail = Some(new_node);
    }

    pub fn display(&self) {
        if let None = &self.tail {
            println!("empty");
        }

        if let Some(_) = &self.head {
            println!("Node count {}", &self.length);
        }
    }
}

fn main() {
    let mut a = LinkedList::new_empty();
    a.append(43);
    a.display();

    // let mut a = Option::Some(45);
    // let b = a.take();

    // assert_eq!(a, None);
    // assert_eq!(b, Some(45));
}
